pageserver/src/tenant/storage_layer.rs

```diff
@@ -204,23 +204,52 @@ impl Default for ValuesReconstructState {
     }
 }
 
-/// Description of layer to be read - the layer map can turn
-/// this description into the actual layer.
-#[derive(PartialEq, Eq, Hash, Debug, Clone)]
-pub(crate) enum ReadableLayerDesc {
-    Persistent {
-        desc: PersistentLayerDesc,
-        lsn_range: Range<Lsn>,
-    },
-    InMemory {
-        handle: InMemoryLayerHandle,
-        lsn_ceil: Lsn,
-    },
+/// A key that uniquely identifies a layer in a timeline
+#[derive(Debug, PartialEq, Eq, Clone, Hash)]
+pub(crate) enum LayerKey {
+    PersitentLayerKey(PersistentLayerKey),
+    InMemoryLayerKey(InMemoryLayerKey),
 }
 
-/// Wraper for 'ReadableLayerDesc' sorted by Lsn
+/// Layer wrapper for the read path. Note that it is valid
+/// to use these layers even after external operations have
+/// been performed on them (compaction, freeze, etc.).
 #[derive(Debug)]
-struct ReadableLayerDescOrdered(ReadableLayerDesc);
+pub(crate) enum ReadableLayer {
+    PersistentLayer(Layer),
+    InMemoryLayer(Arc<InMemoryLayer>),
+}
+
+#[derive(Debug, Clone)]
+pub(crate) struct ReadDesc {
+    layer_key: LayerKey,
+    lsn_range: Range<Lsn>,
+}
+
+impl Ord for ReadDesc {
+    fn cmp(&self, other: &Self) -> Ordering {
+        let ord = self.lsn_range.end.cmp(&other.lsn_range.end);
+        if ord == std::cmp::Ordering::Equal {
+            self.lsn_range.start.cmp(&other.lsn_range.start).reverse()
+        } else {
+            ord
+        }
+    }
+}
+
+impl PartialOrd for ReadDesc {
+    fn partial_cmp(&self, other: &Self) -> Option<Ordering> {
+        Some(self.cmp(other))
+    }
+}
+
+impl PartialEq for ReadDesc {
+    fn eq(&self, other: &Self) -> bool {
+        self.lsn_range == other.lsn_range
```

Comment:
I am surprised the PartialEq does not include the key at all. Are these types we should be putting as `pub(crate)` vs. within some smaller more specialized scope?

Reply:
It doesn't include the key because it should only be used for ordering within the heap. Once the read has been selected, the layer key is then used to index into the second level container.

Also, yes, `ReadDesc` can be private. Will update.

Reply:
Made `ReadDesc` private in https://github.com/neondatabase/neon/pull/7215/commits/6290e5a75cba0da3666dd660596c1d2c97e508d8